### PR TITLE
nixos/stage-1-systemd: ZFS support

### PR DESCRIPTION
**Pull request by ElvishJerricco**
###### Description of changes

Support ZFS when `boot.initrd.systemd.enable` is set to `true`. The bulk of this commit is just moving the `createImportService` function to a higher scope so that it can be reused for the stage 1 systemd units. This also gets us credential requests for free.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/tasks/filesystems/zfs.nix [hunk 1]
```diff
@@ -78,6 +85,80 @@ let
...
+  getPoolMounts = prefix: pool:
+    let
+      mountPoint = fs: escapeSystemdPath (prefix + fs.mountPoint);
+    in
+      map (x: "${mountPoint x}.mount") (getPoolFilesystems pool);
+
+  createImportService = { pool, systemd, force, prefix ? "" }:
+    nameValuePair "zfs-import-${pool}" {
+      description = "Import ZFS pool \"${pool}\"";
+      # we need systemd-udev-settle until https://github.com/zfsonlinux/zfs/pull/4943 is merged
```
Comment (lovesegfault) [inline nixos/modules/tasks/filesystems/zfs.nix:100]: The linked PR is closed, might be good to update this comment.

File: nixos/modules/tasks/filesystems/zfs.nix [hunk 2]
```diff
@@ -78,6 +85,80 @@ let
...
+      environment.ZFS_FORCE = optionalString force "-f";
+      script = (importLib {
+        # See comments at importLib definition.
+        zpoolCmd = "${cfgZfs.package}/sbin/zpool";
+        awkCmd = "${pkgs.gawk}/bin/awk";
+        inherit cfgZfs;
+      }) + ''
+        poolImported "${pool}" && exit
+        echo -n "importing ZFS pool \"${pool}\"..."
+        # Loop across the import until it succeeds, because the devices needed may not be discovered yet.
```
Comment (lovesegfault) [inline nixos/modules/tasks/filesystems/zfs.nix:125]: Why is this needed? I thought it's what udev-settle was for?

**Review by lovesegfault — APPROVED** [review]
LGTM

I left some questions on the code that got moved around. It's also a gnarly script, might be nice to move it to a separate file and so it's easier to shellcheck it. Or maybe use `writeShellApplication`.

File: nixos/modules/tasks/filesystems/zfs.nix (hunk 1 above)
Comment (ElvishJerricco) [reply nixos/modules/tasks/filesystems/zfs.nix:100]: Yea. Should probably leave the link, since it would be good for someone to actually finish that PR, but its status should be reflected.

**Comment by ElvishJerricco** [comment]
Before we merge this, I'm going to add some very simple logic to allow 3 tries to enter the password correctly, since I believe the old initrd does that.

**Comment by ElvishJerricco** [comment]
Going to add a test too.

**Comment by lovesegfault** [comment]
This is waiting on the test to land, I assume?

**Comment by ElvishJerricco** [comment]
@lovesegfault Yea I'm ultimately waiting on https://github.com/NixOS/nixpkgs/pull/169113 to get the zfsroot test, but I'm not sure what to do about its `pivot_root` issue.

**Comment by ElvishJerricco** [comment]
@lovesegfault `zfsroot` installer test is now enabled. I also made a small change to stop using grep since that would have required including grep in initrd.

File: nixos/modules/tasks/filesystems/zfs.nix (hunk 2 above)
Comment (Baughn) [reply nixos/modules/tasks/filesystems/zfs.nix:135]: udev-settle ensures that all pending udev events have been processed. It does _not_ trigger any logic to wait for all hardware probing to finish, etc. On PCs this rarely makes an issue, but when I wrote this, it was to deal with a server-class motherboard that wouldn't finish probing HDDs (and passing them to the OS) until _a minute and a half after the kernel had booted_.

This is not unusual for servers, as I understand it.

In other words:

- This code is untestable on any hardware you're likely to own. Or which I own, for that matter. The C2750 is far too slow to use now.
- If we get it wrong, then a rare few users will be very sad. These are the exact same users most likely to care about reliability.

File: nixos/modules/tasks/filesystems/zfs.nix [hunk 3]
```diff
@@ -58,6 +58,13 @@ let
   # latter case it makes one last attempt at importing, allowing the system to
   # (eventually) boot even with a degraded pool.
   importLib = {zpoolCmd, awkCmd, cfgZfs}: ''
+    for o in $(cat /proc/cmdline); do
+      case $o in
+        zfs_force|zfs_force=1)
```
Comment (Baughn) [inline nixos/modules/tasks/filesystems/zfs.nix:63]: Nit: If you're going to support more than one option, then might as well allow `zfs_force=y` as well.

File: nixos/modules/tasks/filesystems/zfs.nix [hunk 4]
```diff
@@ -78,6 +85,95 @@ let
...
+
+  getKeyLocations = pool:
+    if isBool cfgZfs.requestEncryptionCredentials
+    then "${cfgZfs.package}/sbin/zfs list -rHo name,keylocation,keystatus ${pool}"
+    else "${cfgZfs.package}/sbin/zfs list -Ho name,keylocation,keystatus ${toString (filter (x: datasetToPool x == pool) cfgZfs.requestEncryptionCredentials)}";
+
+  createImportService = { pool, systemd, force, prefix ? "" }:
+    nameValuePair "zfs-import-${pool}" {
+      description = "Import ZFS pool \"${pool}\"";
+      # we need systemd-udev-settle to ensure devices are available
```
Comment (Baughn) [inline nixos/modules/tasks/filesystems/zfs.nix:108]: Just out of curiosity, what happens without it? Can I assume the device nodes simply aren't created?

File: nixos/modules/tasks/filesystems/zfs.nix [hunk 5]
```diff
@@ -78,6 +85,95 @@ let
...
+        # See comments at importLib definition.
+        zpoolCmd = "${cfgZfs.package}/sbin/zpool";
+        awkCmd = "${pkgs.gawk}/bin/awk";
+        inherit cfgZfs;
+      }) + ''
+        poolImported "${pool}" && exit
+        echo -n "importing ZFS pool \"${pool}\"..."
+        # Loop across the import until it succeeds, because the devices needed may not be discovered yet.
+        for trial in `seq 1 60`; do
+          poolReady "${pool}" && poolImport "${pool}" && break
```
Comment (Baughn) [inline nixos/modules/tasks/filesystems/zfs.nix:137]: Since I'm telling stories, note that this only works (on the C2750) because `zpool import` takes a while. It definitely doesn't "wait for a minute at most" here. That's probably fine though.

I always intended to get back and rewrite this in a saner language. Never happened, but I could easily be bribed into doing so now if you think it's a good idea; one virtual cookie will do.

The shell code certainly takes less space than even the most minimalist Rust binary would, of course...

**Review by Baughn — APPROVED** [review]
Seems generally sane.

File: nixos/modules/tasks/filesystems/zfs.nix (hunk 3 above)
Comment (ElvishJerricco) [reply nixos/modules/tasks/filesystems/zfs.nix]: Fair enough

File: nixos/modules/tasks/filesystems/zfs.nix (hunk 4 above)
Comment (ElvishJerricco) [reply nixos/modules/tasks/filesystems/zfs.nix:108]: Not sure to be honest. I've been to afraid to try without it :P
